Quarkus CLI 2.0

patriot1burke opened the pull request:
WIP: I'm still working on testing.

This mostly a port of the aesh/ stuff to Picocli.  I did change command structure a bit here and there.  One specific one is that "list" is renamed to "extensions".  I want the "list" command to list extensions and dependencies that were added to the project with the "add" command.

I'll add some additional comments to this PR that show usage.

patriot1burke committed:
quarkus cli 2.0

cli 2.0 list and rename flags

cli 2.0 version and list usage

cli 2.0 sort = false list options

cli 2.0 PATTERN

cli 2.0 remove aesh from bom

finish cli 2.0 tests

fix compilation error

cli 2.0 fix colors and gradle

cli 2.0 formatting

why is cli test failing in CI

cli try this to get testsuite to pass

cli try yet again to get testsuite to pass

cli maybe this will fix build

cli add some gradle tests

cli revisions

cli width and build refactor

cli fix versoin

cli native build

cli fix

cli --empty to --no-examples

patriot1burke commented:
<pre>Usage: <b>quarkus</b> [<font color="#AA5500">-ehV</font>] [COMMAND]
  <font color="#AA5500">-e</font>, <font color="#AA5500">--errors</font>    Produce execution error messages.
  <font color="#AA5500">-h</font>, <font color="#AA5500">--help</font>      Show this help message and exit.
  <font color="#AA5500">-V</font>, <font color="#AA5500">--version</font>   Print version information and exit.
Commands:
  <b>build</b>             Build your quarkus project
  <b>clean</b>             Clean current project
  <b>create</b>            Create a new quarkus project.
  <b>add</b>               Add extension(s) to current project.
  <b>remove</b>, <b>rm</b>        Remove an extension from this project.
  <b>dev</b>               Execute project in live coding dev mode
  <b>extensions</b>, <b>exts</b>  Display a list of available Quarkus extensions</pre>

patriot1burke commented:
<pre>Usage: <b>quarkus create</b> [<font color="#AA5500">-0hV</font>] [<font color="#AA5500">-a</font>=<i>AID</i>] [<font color="#AA5500">-g</font>=<i>GID</i>] [<font color="#AA5500">-v</font>=<i>TAG</i>] [<font color="#AA5500">--maven</font> | <font color="#AA5500">--gradle</font> |
                      <font color="#AA5500">--grade-kotlin-dsl</font>] [<font color="#AA5500">--java</font> | <font color="#AA5500">--kotlin</font> | <font color="#AA5500">--scala</font>]
                      [<font color="#AA5500">EXTENSION</font>...]
Create a new quarkus project.
      [<font color="#AA5500">EXTENSION</font>...]       extension to add to project
  <font color="#AA5500">-h</font>, <font color="#AA5500">--help</font>               Show this help message and exit.
  <font color="#AA5500">-V</font>, <font color="#AA5500">--version</font>            Print version information and exit.
  <font color="#AA5500">-g</font>, <font color="#AA5500">--groupId</font>=<i>GID</i>        The groupId for project
  <font color="#AA5500">-a</font>, <font color="#AA5500">--artifactId</font>=<i>AID</i>     The artifactId for project
  <font color="#AA5500">-v</font>, <font color="#AA5500">--proj-version</font>=<i>TAG</i>   The version for project
  <font color="#AA5500">-0</font>, <font color="#AA5500">--skipExample</font>        Generate without example code.
<font color="#AA5500"> </font>     <font color="#AA5500">--maven</font>              Create a Maven project. (default)
<font color="#AA5500"> </font>     <font color="#AA5500">--gradle</font>             Create a Gradle project.
<font color="#AA5500"> </font>     <font color="#AA5500">--grade-kotlin-dsl</font>   Create a Gradle Kotlin DSL project.
<font color="#AA5500"> </font>     <font color="#AA5500">--java</font>               Generate Java examples. (default)
<font color="#AA5500"> </font>     <font color="#AA5500">--kotlin</font>             Generate Kotlin examples.
<font color="#AA5500"> </font>     <font color="#AA5500">--scala</font>              Generate Scala examples.
</pre>

patriot1burke commented:
@maxandersen Not sure who else to add to review.

gastaldi commented on devtools/cli/src/main/java/io/quarkus/cli/Create.java:
```diff
@@ -0,0 +1,128 @@
+            }
+
+            SourceType sourceType = SourceType.JAVA;
+            if (targetBuildTool.gradleKotlinDsl)
+                sourceType = SourceType.KOTLIN;
+            else if (language.scala)
+                sourceType = SourceType.SCALA;
+            else if (language.kotlin)
+                sourceType = SourceType.KOTLIN;
+            else if (extensions != null && !extensions.isEmpty())
+                sourceType = CreateProject.determineSourceType(extensions);
+
+            BuildTool buildTool = BuildTool.MAVEN;
+            if (targetBuildTool.gradle)
+                buildTool = BuildTool.GRADLE;
+            else if (targetBuildTool.gradleKotlinDsl)
+                buildTool = BuildTool.GRADLE_KOTLIN_DSL;
```
Looks like brackets are no longer a thing in 2020 :)

patriot1burke replied:
Seriously?

gastaldi replied:
Well, yeah :) IMHO it looks better with brackets :)

gastaldi replied:
It's all fun and games until someone adds a new line and wonders why it doesn't work as it should :wink:

patriot1burke replied:
I just don't agree. :) We're not coding in 'vi' or 'ed' anymore.

gsmet commented:
Minor nitpicks: I think you should be consistent for the long options casing:
- `groupId` -> `group-id`
- `artifactId` -> `artifact-id`
- `proj-version` -> `project-version`
- `skipExample` -> `skip-example` (I would even prefer `skip-examples` as I think you might have several) @ia3andy I think we should change it in the code starts code if I'm right. Plural is more natural in English.

I would drop the current `--version` from `create` and replace `--project-version` with it.

`quarkus -v` displaying the version makes sense.
`quarkus create -v` displaying the version makes less sense IMHO.

gastaldi commented:
I think `aesh` can be removed from the `bom/application/pom.xml` too?

patriot1burke commented:
> Minor nitpicks: I think you should be consistent for the long options casing:
> 
> * `groupId` -> `group-id`
> * `artifactId` -> `artifact-id`
> * `proj-version` -> `project-version`
> * `skipExample` -> `skip-example` (I would even prefer `skip-examples` as I think you might have several) @ia3andy I think we should change it in the code starts code if I'm right. Plural is more natural in English.
> 
> I would drop the current `--version` from `create` and replace `--project-version` with it.
> 
> `quarkus -v` displaying the version makes sense.
> `quarkus create -v` displaying the version makes less sense IMHO.

This is gonna sound stupid, but i renamed from project-version to proj-version for Usage formatting reasons so I could keep the 80 char width.

gsmet commented:
> This is gonna sound stupid, but i renamed from project-version to proj-version for Usage formatting reasons so I could keep the 80 char width.

An even better reason to use `--version` (if possible).

patriot1burke commented:
> Minor nitpicks: I think you should be consistent for the long options casing:
> 
> * `groupId` -> `group-id`
> * `artifactId` -> `artifact-id`
> * `proj-version` -> `project-version`
> * `skipExample` -> `skip-example` (I would even prefer `skip-examples` as I think you might have several) @ia3andy I think we should change it in the code starts code if I'm right. Plural is more natural in English.
> 
> I would drop the current `--version` from `create` and replace `--project-version` with it.
> 
> `quarkus -v` displaying the version makes sense.
> `quarkus create -v` displaying the version makes less sense IMHO.

I made the flags consistent with quarkus:XXXX where it made sense.  Either we're keeping the flags consistent  or we aren't.  CHOOSE!

gsmet commented:
Yeah, the `skipExample` one, I think we should change in quarkus:xxx too. But we need to check with @ia3andy if it makes sense.

As for `version`, I think it's just that it's probably something we couldn't use with Maven. Here it makes perfect sense to use it if we can.

If we can make things better with this CLI, let's make them, I think the idea is to have people use it rather than the Maven counterpart.

patriot1burke commented:
> I made the flags consistent with quarkus:XXXX where it made sense. Either we're keeping the flags consistent or we aren't. CHOOSE!

For this I meant whether kebel case should be used or not "artifactId" vs. "artifact-id".

gastaldi commented:
+1 to kebab case everywhere. That's the same pattern used in the JSON files too

patriot1burke commented:
<pre>Usage: <b>quarkus list</b> [<font color="#AA5500">-hi</font>] [<font color="#AA5500">-s</font>=<i>PATTERN</i>] [<font color="#AA5500">--name</font> | <font color="#AA5500">--concise</font> | <font color="#AA5500">--full</font>]
List installed (default) or installable extensions.
  <font color="#AA5500">-h</font>, <font color="#AA5500">--help</font>             display this help message
  <font color="#AA5500">-i</font>, <font color="#AA5500">--installable</font>      Display installable extensions.
  <font color="#AA5500">-s</font>, <font color="#AA5500">--search</font>=<i>PATTERN</i>   Search filter on extension list. The format is based
                           on Java Pattern.
<font color="#AA5500"> </font>     <font color="#AA5500">--name</font>             Display extension name only. (default)
<font color="#AA5500"> </font>     <font color="#AA5500">--concise</font>          Display extension name and description.
<font color="#AA5500"> </font>     <font color="#AA5500">--full</font>             Display concise format and version related columns.
</pre>

patriot1burke commented:
Ok, see updated usage messages above base, create, and list.

I removed "extensions" command and replaced it back to "list".  By default "list" will list installed extensions.  "--installable" will list installable extensions.  I also removed descriptor text like "Current extensions blah blah blah" as you might want to pipe this stuff in a script.

ia3andy commented:
Regarding `skip-examples`, I will change it in codestarts/maven too, but I would personally prefer just: `—empty` as we discussed on quarkus-dev.

patriot1burke commented:
@ia3andy changed to --empty

Added tests.  Everything should be good to go.

ia3andy commented on devtools/cli/src/main/java/io/quarkus/cli/Add.java:
```diff
@@ -0,0 +1,53 @@
+    public Integer call() {
+        Path projectDirectory = Paths.get(System.getProperty("user.dir"));
+        try {
+            QuarkusProject quarkusProject = QuarkusProject.resolveExistingProject(projectDirectory,
```
@aloubyansky @maxandersen @patriot1burke , this command is not going to be compatible with gradle. This is why for the add/remove/list extension, I would delegate the call to the wrapper plugin to make this command compatible on both mvn and gradle. If there is no wrapper, then you could fallback to this and if it target a gradle project fail with instruction.

ia3andy replied:
And even more now that I see you already implemented an `ExecuteUtil` :)

patriot1burke replied:
Can you explain why it won't work?

ia3andy replied:
sure @patriot1burke, because Gradle API is not available outside of a gradle plugin, without it, we are not able to properly list extensions (neither add and remove which rely on the list): https://github.com/quarkusio/quarkus/blob/master/independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/buildfile/GenericGradleBuildFile.java

maxandersen replied:
can confirm it doesn't work:

```
 quarkus add hibernate-orm-panache
Unable to add extension: This feature is not yet implemented outside of the Gradle Plugin.
```

ia3andy commented on devtools/cli/src/main/java/io/quarkus/cli/Create.java:
```diff
@@ -0,0 +1,131 @@
+                            .groupId(groupId)
+                            .artifactId(artifactId)
+                            .version(version)
+                            .sourceType(sourceType)
```
@patriot1burke fyi `sourceType` is not used by codestarts, you could remove it. What you will need to do though, is to add the language extension when `--kotlin` or `--scala` is selected.

patriot1burke replied:
Don't know what "add the language extension" means...

ia3andy replied:
@patriot1burke:
- java -> default
- kotlin -> extension `quarkus-kotlin`
- scala -> extension `quarkus-scala`

in the end the `--kotlin` or `--scala` are just shortcuts.

ia3andy replied:
yes!

ia3andy reviewed (changes requested):
Very nice! Just a comment about add/remove/list commands..

maxandersen commented:
First off - maan this is awesome to have kicked off; it feels so much more natural and complete :) ...will be adding some comments for some issues I found and comments!

maxandersen commented on devtools/cli/src/main/java/io/quarkus/cli/Remove.java:
```diff
@@ -0,0 +1,46 @@
+        } catch (Exception e) {
+            if (parent.showErrors)
+                e.printStackTrace(spec.commandLine().getErr());
+            spec.commandLine().getErr().println("Unable to remove extension matching:" + e.getMessage());
```
`quarkus rm hibernate` in a project with just `quarkus-hibernate-orm-panache` still complains about having to be precise:

```
❌ Multiple extensions matching 'hibernate'
     * io.quarkus:quarkus-hibernate-orm
     * io.quarkus:quarkus-hibernate-reactive
     * io.quarkus:quarkus-hibernate-envers
     * io.quarkus:quarkus-hibernate-orm-panache
     * io.quarkus:quarkus-hibernate-orm-rest-data-panache
     * io.quarkus:quarkus-hibernate-orm-panache-kotlin
     * io.quarkus:quarkus-hibernate-search-elasticsearch
     * io.quarkus:quarkus-hibernate-validator
     Be more specific e.g using the exact name or the full GAV.
```

patriot1burke replied:
This should be logged as a separate issue.  Not a CLI thing.

maxandersen commented on devtools/cli/src/main/java/io/quarkus/cli/core/ExecuteUtil.java:
```diff
@@ -0,0 +1,159 @@
+
+        InvocationResult result = invoker.execute(request);
+
+        if (result.getExitCode() != 0) {
```
when no no `mvnw` nor `mvn` in path this seem to just fail silently. no error output, no nothing.

patriot1burke replied:
stderr wasn't being outputted.  All this code has changed and this is fixed.

maxandersen commented on devtools/cli/pom.xml:
```diff
@@ -0,0 +1,119 @@
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>build</goal>
```
Tried building with `-Dnative` and I get errors; at least on OSX.

```
Error: Unsupported features in 3 methods
Detailed message:
Error: Class initialization of io.netty.util.internal.logging.Log4JLogger failed. Use the option --initialize-at-run-time=io.netty.util.internal.logging.Log4JLogger to explicitly request delayed initialization of this class.
Original exception that caused the problem: java.lang.NoClassDefFoundError: org/apache/log4j/Priority
```

patriot1burke replied:
added --native, -n switch.

patriot1burke replied:
added --native switch to build

maxandersen commented on devtools/cli/src/main/java/io/quarkus/cli/Build.java:
```diff
@@ -0,0 +1,29 @@
+    boolean usageHelpRequested;
+
+    @CommandLine.ParentCommand
+    QuarkusCli parent;
```
all this parent/spec code and possibly also the error handling we could move to a base class and simplify these commands quite a bit.

maxandersen commented on devtools/cli/src/main/java/io/quarkus/cli/Build.java:
```diff
@@ -0,0 +1,29 @@
+
+@CommandLine.Command(name = "build", mixinStandardHelpOptions = false, description = "Build your quarkus project")
+public class Build implements Callable<Integer> {
+    @CommandLine.Option(names = { "-h", "--help" }, usageHelp = true, description = "display this help message")
```
should we have a `--verbose` that would print out the commands actually executed ?

maxandersen commented on devtools/cli/src/main/java/io/quarkus/cli/Build.java:
```diff
@@ -0,0 +1,29 @@
+import io.quarkus.cli.core.ExecuteUtil;
+import picocli.CommandLine;
+
+@CommandLine.Command(name = "build", mixinStandardHelpOptions = false, description = "Build your quarkus project")
```
looks like I can do `quarkus clean build` and that gets execute as multiple commands. That makes it slower than corresponding `mvn clean build` or `gradle clean build`

Wondering if we could "group" these rather than be done separately ?

maxandersen replied:
haven't played with it but picocli lets you customize the execution strategy. https://picocli.info/#_executing_commands_with_subcommands

maybe that could work ?

patriot1burke replied:
creating a strategy was the trick.

patriot1burke replied:
I used a strategy.  That worked.

maxandersen reviewed (changes requested):
overall this looks awesome IMO. 

My comments are probably for most part things we can do in follow up PR's; only "glitch" I think should be fixed before merging is that "add"/"remove" aren't working on gradle projects.

ia3andy commented on devtools/cli/src/main/java/io/quarkus/cli/Add.java:
```diff
@@ -0,0 +1,80 @@
+    private Integer addGradle(Path projectDirectory) {
+        ArrayList<String> args = new ArrayList<>();
+        args.add("addExtension");
+        String param = "--extensions=";
+        boolean first = true;
+        for (String ext : extensions) {
+            if (first)
+                first = false;
+            else
+                param += ",";
+
+            param += ext;
+        }
```
:-)

patriot1burke commented on devtools/cli/src/main/java/io/quarkus/cli/Create.java:
```diff
@@ -0,0 +1,131 @@
+package io.quarkus.cli;
```
How is that simpler than:
$ quarkus create resteasy

???

patriot1burke replied:
Create a separate issue if you want this please.

maxandersen replied:
create takes extensions as arguments not a codestart thus you can end up with a mix of things in your project. Which is good when exploring; less great when trying to write docs or follow a guide that just want "take this starting point, add these extensions and we'll walk you through the rest".

i'll open seperate issue.

patriot1burke commented:
<pre>Usage: <b>quarkus</b> [<font color="#AA5500">-ehV</font>] [<font color="#AA5500">--verbose</font>] [COMMAND]
  <font color="#AA5500">-e</font>, <font color="#AA5500">--errors</font>    Produce execution error messages.
  <font color="#AA5500">-h</font>, <font color="#AA5500">--help</font>      Show this help message and exit.
  <font color="#AA5500">-V</font>, <font color="#AA5500">--version</font>   Print version information and exit.
      <font color="#AA5500">--verbose</font>   Verbose mode.
Commands:
  <b>build</b>       Build your quarkus project
  <b>clean</b>       Clean current project
  <b>create</b>      Create a new quarkus project.
  <b>list</b>        List installed (default) or installable extensions.
  <b>add</b>         Add extension(s) to current project.
  <b>remove</b>, <b>rm</b>  Remove an extension from this project.
  <b>dev</b>         Execute project in live coding dev mode
</pre>

patriot1burke commented:
<pre>Usage: <b>quarkus build</b> [<font color="#AA5500">-hns</font>] [<font color="#AA5500">--offline</font>]
Build your quarkus project
  <font color="#AA5500">-h</font>, <font color="#AA5500">--help</font>         display this help message
  <font color="#AA5500">-n</font>, <font color="#AA5500">--native</font>       Build native executable.
      <font color="#AA5500">--offline</font>      Work offline.
  <font color="#AA5500">-s</font>, <font color="#AA5500">--skip-tests</font>   Skip tests.
</pre>

patriot1burke commented:
<pre>Usage: <b>quarkus create</b> [<font color="#AA5500">-0h</font>] [<font color="#AA5500">-a</font>=<i>ARTIFACT-ID</i>] [<font color="#AA5500">-g</font>=<i>GROUP-ID</i>] [<font color="#AA5500">-v</font>=<i>VERSION</i>] [<font color="#AA5500">--maven</font> | <font color="#AA5500">--gradle</font> |
                      <font color="#AA5500">--grade-kotlin-dsl</font>] [<font color="#AA5500">--java</font> | <font color="#AA5500">--kotlin</font> | <font color="#AA5500">--scala</font>] [<font color="#AA5500">EXTENSION</font>...]
Create a new quarkus project.
      [<font color="#AA5500">EXTENSION</font>...]              extension to add to project
  <font color="#AA5500">-h</font>, <font color="#AA5500">--help</font>                      display this help message
  <font color="#AA5500">-g</font>, <font color="#AA5500">--group-id</font>=<i>GROUP-ID</i>         The groupId for project
  <font color="#AA5500">-a</font>, <font color="#AA5500">--artifact-id</font>=<i>ARTIFACT-ID</i>   The artifactId for project
  <font color="#AA5500">-v</font>, <font color="#AA5500">--version</font>=<i>VERSION</i>           The version for project
  <font color="#AA5500">-0</font>, <font color="#AA5500">--empty</font>                     Generate without example code.
<font color="#AA5500"> </font>     <font color="#AA5500">--maven</font>                     Create a Maven project. (default)
<font color="#AA5500"> </font>     <font color="#AA5500">--gradle</font>                    Create a Gradle project.
<font color="#AA5500"> </font>     <font color="#AA5500">--grade-kotlin-dsl</font>          Create a Gradle Kotlin DSL project.
<font color="#AA5500"> </font>     <font color="#AA5500">--java</font>                      Generate Java examples. (default)
<font color="#AA5500"> </font>     <font color="#AA5500">--kotlin</font>                    Generate Kotlin examples.
<font color="#AA5500"> </font>     <font color="#AA5500">--scala</font>                     Generate Scala examples.
</pre>

patriot1burke commented:
<pre>List installed (default) or installable extensions.
  <font color="#AA5500">-h</font>, <font color="#AA5500">--help</font>             display this help message
  <font color="#AA5500">-i</font>, <font color="#AA5500">--installable</font>      Display installable extensions.
  <font color="#AA5500">-s</font>, <font color="#AA5500">--search</font>=<i>PATTERN</i>   Search filter on extension list. The format is based on Java Pattern.
<font color="#AA5500"> </font>     <font color="#AA5500">--name</font>             Display extension name only. (default)
<font color="#AA5500"> </font>     <font color="#AA5500">--concise</font>          Display extension name and description.
<font color="#AA5500"> </font>     <font color="#AA5500">--full</font>             Display concise format and version related columns.
</pre>

patriot1burke commented:
@maxandersen I think I addressed all issues and added some more functionality.  See above for updated usage.  Assuming I can get CI to pass, this should be good to go.  That's a big IF though...I've been unable to get CI to pass this whole last week even though I can get it to pass locally on windows and linux.

patriot1burke commented:
@maxandersen Yay a passed build.  If you think everything is cool an approval would be welcome.

patriot1burke commented:
@maxandersen native is going to have to come with a different PR.

maxandersen commented:
Awesome - will look when back over weekend or first Monday!

maxandersen commented:
I think this is good to go. still things needed but as we aren't shipping any bits yet using this its not hurting to go in and we can get started on the bigger things.

LGTM!

maxandersen reviewed (approved):
LGTM!

ia3andy commented on devtools/cli/src/main/java/io/quarkus/cli/Create.java:
```diff
@@ -0,0 +1,140 @@
+            if (targetBuildTool.gradleKotlinDsl) {
+                sourceType = SourceType.KOTLIN;
+                if (extensions == null)
+                    extensions = new HashSet<>();
```
you could do it only once I guess

patriot1burke replied:
it saves an object creation when its --java and no extensions.

ia3andy commented on devtools/cli/src/main/java/io/quarkus/cli/Create.java:
```diff
@@ -0,0 +1,140 @@
+
+    @CommandLine.Option(names = { "-0",
+            "--empty" }, order = 4, description = "Generate without example code.")
+    boolean empty = false;
```
@maxandersen didn't we say we should use `no-examples` in the end?

patriot1burke replied:
Fine to change this, but I'm gonna kill somebody if I have to spend another week trying to get CI to pass over such a small change.

ia3andy replied:
@patriot1burke I can understand this 😂, it shouldn't break CI though :) let's wait for @maxandersen confirmation first.. 

This is not "that" small because it actually changes the spec of the CLI.. ex: if you give an example in a blog post and it changes in the next release, it's not great 😅

patriot1burke replied:
Well, you guys have already changed this flag 3 times in the last two weeks and , well, i don't like any of your changes :)  So, if you change your mind yet again in the next two weeks, feel free to do the change yourself.

maxandersen replied:
i wasn't aware we were changing them here as those things were stil fluid.

lets just get this in and we can clean up / align these rather than do another round.

patriot1burke commented:
@maxandersen @ia3andy FYI: Maybe i just don't have gradle set up correctly, but any command (build, clean, list, add, remove) that goes through gradle is painfully slow.

maxandersen commented:
lets get this into master - it just missed the 1.8 cut off ;/

ia3andy reviewed (approved):
LGTM
